File: Sources/BsnsSSHCore/SSHWire.swift
import Foundation

/// Minimal SSH wire-format encoder (RFC 4251 §5). Only the pieces the agent
/// and signature framing actually need: `uint32`, `string`, and the
/// sign-padded `mpint`.
///
/// `mpint` is the one with a footgun: it is a two's-complement big-endian
/// integer stored as a string, so a non-negative value whose top bit is set
/// must carry a leading 0x00 or it reads as negative — and unnecessary leading
/// zero bytes are forbidden. Getting this wrong makes public-key auth fail
/// silently, so it is covered by the RFC reference vectors in the tests.
public struct SSHEncoder {
    public private(set) var data = Data()

    public init() {}

    /// Append a 32-bit big-endian integer.
    public mutating func writeUInt32(_ value: UInt32) {
        var be = value.bigEndian
        withUnsafeBytes(of: &be) { data.append(contentsOf: $0) }
    }

    /// Append raw bytes with no length prefix.
    public mutating func writeBytes(_ bytes: Data) {
        data.append(bytes)
    }

    /// SSH `string`: a `uint32` length followed by that many bytes.
    public mutating func writeString(_ bytes: Data) {
        writeUInt32(UInt32(bytes.count))
        data.append(bytes)
    }

    /// SSH `string` from UTF-8 text.
    public mutating func writeString(_ string: String) {
        writeString(Data(string.utf8))
    }

    /// SSH `mpint`: a non-negative big-endian magnitude encoded as a string,
    /// with leading zero bytes stripped and a single 0x00 prepended when the
    /// high bit of the leading byte is set. Zero encodes as an empty string.
    public mutating func writeMPInt(_ magnitude: Data) {
        var bytes = [UInt8](magnitude)
        while bytes.first == 0x00 { bytes.removeFirst() }
        if bytes.isEmpty {
            writeUInt32(0)
            return
        }
        if bytes[0] & 0x80 != 0 {
            bytes.insert(0x00, at: 0)
        }
        writeString(Data(bytes))
    }
}

public extension SSHEncoder {
    /// Build a blob inline and return its bytes.
    static func build(_ body: (inout SSHEncoder) -> Void) -> Data {
        var encoder = SSHEncoder()
        body(&encoder)
        return encoder.data
    }
}

File: android/core/src/main/kotlin/cc/bsns/ssh/core/SshWire.kt
package cc.bsns.ssh.core

/**
 * Minimal SSH wire-format encoder (RFC 4251 §5) — the Kotlin port of the Swift
 * `SSHEncoder`. Only the pieces the agent and signature framing need: `uint32`,
 * `string`, and the sign-padded `mpint`.
 *
 * This MUST stay byte-identical to the Swift implementation; the parity test
 * vectors (shared with the iOS `SSHWireTests`) are the contract. `mpint` is the
 * footgun: a non-negative big-endian magnitude stored as a string, leading zero
 * bytes stripped, a single 0x00 prepended when the high bit of the leading byte
 * is set, and zero encoded as an empty string.
 */
class SshEncoder {
    private val buf = ArrayList<Byte>()

    val data: ByteArray get() = buf.toByteArray()

    fun writeByte(value: Int) {
        buf.add(value.toByte())
    }

    /** 32-bit big-endian. Takes a Long so the full unsigned range is expressible. */
    fun writeUInt32(value: Long) {
        buf.add(((value ushr 24) and 0xFF).toByte())
        buf.add(((value ushr 16) and 0xFF).toByte())
        buf.add(((value ushr 8) and 0xFF).toByte())
        buf.add((value and 0xFF).toByte())
    }

    fun writeBytes(bytes: ByteArray) {
        for (b in bytes) buf.add(b)
    }

    /** SSH `string`: a uint32 length followed by that many bytes. */
    fun writeString(bytes: ByteArray) {
        writeUInt32(bytes.size.toLong())
        writeBytes(bytes)
    }

    fun writeString(text: String) = writeString(text.toByteArray(Charsets.UTF_8))

    /** SSH `mpint` — see the class doc for the sign-padding rule. */
    fun writeMPInt(magnitude: ByteArray) {
        var start = 0
        while (start < magnitude.size && magnitude[start] == 0x00.toByte()) start++
        if (start == magnitude.size) {           // empty or all-zero → zero
            writeUInt32(0)
            return
        }
        val stripped = magnitude.copyOfRange(start, magnitude.size)
        val out = if (stripped[0].toInt() and 0x80 != 0) byteArrayOf(0x00) + stripped else stripped
        writeString(out)
    }

    companion object {
        /** Build a blob inline and return its bytes. */
        fun build(body: (SshEncoder) -> Unit): ByteArray {
            val e = SshEncoder()
            body(e)
            return e.data
        }
    }
}
